app/src/main/java/ch/epfl/balelecbud/view/gallery/GalleryFragment.java

```diff
@@ -1,48 +1,149 @@
 import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
 
+import com.google.android.material.floatingactionbutton.FloatingActionButton;
+import com.google.android.material.snackbar.Snackbar;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Date;
+
 import ch.epfl.balelecbud.R;
 import ch.epfl.balelecbud.model.Picture;
+import ch.epfl.balelecbud.utility.DateFormatter;
 import ch.epfl.balelecbud.utility.recyclerViews.RefreshableRecyclerViewAdapter;
 
+import static ch.epfl.balelecbud.BalelecbudApplication.getAppStorage;
+import static ch.epfl.balelecbud.utility.storage.Storage.USER_PICTURES;
+import static com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_SHORT;
+
 public final class GalleryFragment extends Fragment {
 
     public final static String TAG = GalleryFragment.class.getSimpleName();
+    private static final int CAMERA_PERM_CODE = 101;
+    private static final int CAMERA_REQUEST_CODE = 102;
+    private RefreshableRecyclerViewAdapter<Picture, PictureHolder> adapter;
+    private String currentPhotoPath;
 
     public static GalleryFragment newInstance() {
         return new GalleryFragment();
     }
 
     @Override
-    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         return inflater.inflate(R.layout.fragment_gallery, container, false);
     }
 
     @Override
-    public void onStart(){
+    public void onStart() {
         super.onStart();
         RecyclerView rvGallery = getView().findViewById(R.id.gallery_recycler_view);
-        View freshnessView = getView().findViewById(R.id.freshness_info_layout);
-
 
         PictureData data = new PictureData();
-        RefreshableRecyclerViewAdapter<Picture, PictureHolder> adapter = new RefreshableRecyclerViewAdapter<>(
-                PictureHolder::new, freshnessView, data, R.layout.item_gallery);
+        adapter = new RefreshableRecyclerViewAdapter<>(PictureHolder::new, rvGallery, data, R.layout.item_gallery);
 
         rvGallery.setLayoutManager(new GridLayoutManager(getActivity(), 3));
         rvGallery.setAdapter(adapter);
-
         SwipeRefreshLayout refreshLayout = getActivity().findViewById(R.id.swipe_refresh_layout_gallery);
         adapter.setOnRefreshListener(refreshLayout);
+
+        FloatingActionButton cameraButton = getActivity().findViewById(R.id.fab_upload_picture);
+        cameraButton.setOnClickListener(v -> askCameraPermissions());
+
+    }
+
+    private void askCameraPermissions() {
+        if (ContextCompat.checkSelfPermission(getActivity(), Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(getActivity(), new String[]{Manifest.permission.CAMERA}, CAMERA_PERM_CODE);
+        } else {
+            dispatchTakePictureIntent();
+        }
+
```

Comment:
You could always request the permission and if you already have it the OS will directly call `onRequestPermissionResult` without prompting the user
```suggestion
        ActivityCompat.requestPermissions(getActivity(), new String[]{Manifest.permission.CAMERA}, CAMERA_PERM_CODE);
```